crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/space.rs

```diff
@@ -225,63 +225,93 @@ impl SpaceInternalHandler for Space {
-        }
-
-        // to_agent *is* joined - let's forward the request
+    ) -> SpaceInternalHandlerResult<wire::Wire> {
         let space = self.space.clone();
+        if self.agents.contains_key(&to_agent) {
+            // LOCAL SHORT CIRCUIT! - just forward data locally
 
-        // clone the event sender
-        let evt_sender = self.evt_sender.clone();
+            let evt_sender = self.evt_sender.clone();
 
-        // As this is a short-circuit - we need to decode the data inline - here.
-        // In the future, we will probably need to branch here, so the real
-        // networking can forward the encoded data. Or, split immediate_request
-        // into two variants, one for short-circuit, and one for real networking.
-        let (_, data) = wire::Wire::decode_ref(&data)?;
+            let (_, data) = wire::Wire::decode_ref(&data)?;
 
-        match data {
-            wire::Wire::Call(payload) => Ok(async move {
-                evt_sender
-                    .call(space, to_agent, from_agent, payload.data.into())
-                    .await
-            }
-            .instrument(tracing::debug_span!("wire_call"))
-            .boxed()
-            .into()),
-            wire::Wire::Notify(payload) => {
-                Ok(async move {
+            match data {
+                wire::Wire::Call(payload) => Ok(async move {
+                    let res = evt_sender
+                        .call(space, to_agent, from_agent, payload.data.into())
+                        .await?;
+                    Ok(wire::Wire::call_resp(res.into()))
+                }
+                .instrument(tracing::debug_span!("wire_call"))
+                .boxed()
+                .into()),
+                wire::Wire::Notify(payload) => Ok(async move {
                     evt_sender
                         .notify(space, to_agent, from_agent, payload.data.into())
                         .await?;
-                    // broadcast doesn't return anything...
-                    Ok(vec![])
+                    Ok(wire::Wire::notify_resp())
                 }
                 .boxed()
-                .into())
+                .into()),
+                _ => unimplemented!(),
             }
-            _ => {
-                tracing::warn!("UNHANDLED WIRE: {:?}", data);
-                Ok(async move { Ok(vec![]) }.boxed().into())
+        } else {
+            let evt_sender = self.evt_sender.clone();
+            let tx = self.transport.clone();
+            Ok(async move {
+                // see if we have an entry for this agent in our agent_store
+                let info = match evt_sender
+                    .get_agent_info_signed(GetAgentInfoSignedEvt {
+                        space,
+                        agent: to_agent.clone(),
+                    })
+                    .await?
+                {
+                    None => return Err(KitsuneP2pError::RoutingAgentError(to_agent)),
```

Comment:
How come this is a routing error? Wouldn't it mean that you just don't have the agent in your peer store because they weren't gossiped to you?

Reply:
This is in `immediate_request`... the current logic has this function either send a message now or error - and the loops above this function manage waiting for us to potentially get more peer info and try again.

This will probably need to change when we write the master peer-discovery logic.